libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs

```diff
@@ -0,0 +1,185 @@
+
+        /// <summary>
+        /// The Azure Storage Blob Container where entities will be stored
+        /// </summary>
+        private Lazy<ValueTask<CloudBlobContainer>> Container { get; set; }
+
+        /// <summary>
+        /// Creates the AzureBlobStorage instance
+        /// </summary>
+        /// <param name="dataConnectionString">Azure Storage connection string</param>
+        /// <param name="containerName">Name of the Blob container where entities will be stored</param>
+        public AzureBlobStorage(string dataConnectionString, string containerName)
+            : this(CloudStorageAccount.Parse(dataConnectionString), containerName)
+        {
+        }
+
+        /// <summary>
+        /// Creates the AzureBlobStorage instance
+        /// </summary>
+        /// <param name="storageAccount">Azure CloudStorageAccount instance</param>
+        /// <param name="containerName">Name of the Blob container where entities will be stored</param>
+        public AzureBlobStorage(CloudStorageAccount storageAccount, string containerName)
+        {
+            if (storageAccount == null) throw new ArgumentNullException(nameof(storageAccount));
+
+            // Checks if a container name is valid
+            NameValidator.ValidateContainerName(containerName);
+
+            this.Container = new Lazy<ValueTask<CloudBlobContainer>>(async () =>
+            {
+                var blobClient = storageAccount.CreateCloudBlobClient();
+                var container = blobClient.GetContainerReference(containerName);
+                await container.CreateIfNotExistsAsync().ConfigureAwait(false);
+                return container;
+            }, isThreadSafe: true);
+        }
+
+        /// <summary>
+        /// Get a blob name validated representation of an entity
+        /// </summary>
+        /// <param name="key">The key used to identify the entity</param>
+        /// <returns></returns>
+        private static string GetBlobName(string key)
+        {
+            if (string.IsNullOrEmpty(key)) throw new ArgumentNullException(nameof(key));
+
+            var blobName = HttpUtility.UrlEncode(key);
+            NameValidator.ValidateBlobName(blobName);
+            return blobName;
+        }
+
+        /// <summary>
+        /// Deletes entity blobs from the configured container
+        /// </summary>
+        /// <param name="keys">An array of entity keys</param>
+        /// <returns></returns>
+        public async Task Delete(string[] keys)
+        {
+            if (keys == null) throw new ArgumentNullException(nameof(keys));
+
+            await Task.WhenAll(
+                keys.Select(async(key) =>
+                {
+                    var blobName = GetBlobName(key);
+                    var blobContainer = await this.Container.Value;
+                    var blobReference = blobContainer.GetBlobReference(blobName);
+                    return blobReference.DeleteIfExistsAsync();

```

Comment:
no await here?  That seems wrong

Reply:
Agree. However, instead we moved the `var blobContainer = await this.Container.Value;` out of the lambda. Now the `Task.WhenAll()` receives the list of `blobReference.DeleteIfExistsAsync()` tasks.
Similar to [this](https://github.com/Microsoft/botbuilder-dotnet/blob/fe0a21fc1fd8580865f0aed4f2fe6a2fc40c12a3/libraries/Microsoft.Bot.Builder.Azure/CosmosDbStorage.cs#L112-L115).